functions/src/aqi-calculation/cleaned-reading.ts

```diff
@@ -1,11 +1,226 @@
+          buffer.length + startIndex,
+          buffer.length + endIndex
+        );
+      }
+      endIndex = startIndex;
+      startIndex -= ELEMENTS_PER_HOUR;
+
+      // If we have 1 reading every two minutes, there are 30 readings in an hour.
+      // 75% of 30 readings is 23 (22.5) readings. As suggested by the EPA, we use
+      // 75% instead of 90% so that sensors are more likely to have enough valid
+      // data. The acceptability of 75% instead of 95% can be found in the additional
+      // slides of the PDF located at the following URL:
+      // https://cfpub.epa.gov/si/si_public_file_download.cfm?p_download_id=540979&Lab=CEMM
+      // Expressed this way to avoid imprecision of floating point arithmetic.
+      const MEASUREMENT_COUNT_THRESHOLD = 23;
+      // Remove all invalid readings
+      readings.filter(element => element.timestamp !== null);
+      if (readings.length >= MEASUREMENT_COUNT_THRESHOLD) {
+        averages[hoursAgo] = averageReadings(readings);
+      }
+    }
   }
+  return averages;
 }
+
+/**
+ * Cleans hourly averages of PM2.5 readings using the published EPA formula,
+ * excluding thoses data points that indicate sensor malfunction. Those
+ * data points are represented by NaN.
+ *
+ * @param averages - array containing sensor readings representing hourly averages
+ * @returns an array of numbers representing the corrected PM2.5 values pursuant to the EPA formula
+ */
+function cleanAverages(averages: BasicReading[]): number[] {
+  // These thresholds for the EPA indicate when diverging sensor readings
+  // indicate malfunction. The EPA requires that the raw difference between
+  // the readings be less than 5 and the percent difference be less than 70%
+  const RAW_THRESHOLD = 5;
+  const PERCENT_THRESHOLD = 0.7;
+
+  const cleanedAverages = new Array<number>(averages.length);
+
+  for (let i = 0; i < cleanedAverages.length; i++) {
+    const reading = averages[i];
+    if (reading) {
+      const averagePmReading =
+        (reading.channelAPm25 + reading.channelBPm25) / 2; // eslint-disable-line no-magic-numbers
+      const difference = Math.abs(reading.channelAPm25 - reading.channelBPm25);
+      if (
+        !(
+          difference > RAW_THRESHOLD &&
+          difference / averagePmReading > PERCENT_THRESHOLD
+        )
+      ) {
+        // Formula from EPA to correct PurpleAir PM2.5 readings
+        // https://cfpub.epa.gov/si/si_public_record_report.cfm?dirEntryId=349513&Lab=CEMM&simplesearch=0&showcriteria=2&sortby=pubDate&timstype=&datebeginpublishedpresented=08/25/2018
+        /* eslint-disable no-magic-numbers */
+        cleanedAverages[i] =
+          0.534 * averagePmReading - 0.0844 * reading.humidity + 5.604;
+        /* eslint-enable no-magic-numbers */
+      } else {
+        // If reading exceeds thresholds above
+        cleanedAverages[i] = Number.NaN;
+      }
+    } else {
+      // If less than 27 data points were available for that hour, the reading
```

Comment:
This comment is now misleading since we are using 75% not 95%, sorry for not catching in the earlier PR where we swtiched